clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3515,4 +3668,198 @@ void handleFailure(Throwable throwable) {
+            }
+        }
+        for (Map.Entry<TopicPartition, OffsetSpec> entry: topicPartitionOffsets.entrySet()) {
+
+            OffsetSpec offsetSpec = entry.getValue();
+            long offsetQuery = (offsetSpec instanceof TimestampSpec)
+                    ? ((TimestampSpec)offsetSpec).timestamp()
+                    : (offsetSpec instanceof OffsetSpec.EarliestSpec)
+                        ? ListOffsetRequest.EARLIEST_TIMESTAMP
+                        : ListOffsetRequest.LATEST_TIMESTAMP;
+            // avoid sending listOffsets request for topics with errors
+            if (!context.getErrors().containsKey(entry.getKey().topic())) {
+                Node node = context.getCluster().leaderFor(entry.getKey());
+                if (node != null) {
+                    if (!leaders.containsKey(node))
+                        leaders.put(node, new HashMap<>());
+                    leaders.get(node).put(entry.getKey(), new ListOffsetRequest.PartitionData(offsetQuery, leaderEpochs.get(entry.getKey())));
+                } else {
+                    KafkaFutureImpl<ListOffsetsResultInfo> future = futures.get(entry.getKey());
+                    future.completeExceptionally(Errors.LEADER_NOT_AVAILABLE.exception());
+                }
+            }
+        }
+
+        for (final Map.Entry<Node, Map<TopicPartition, ListOffsetRequest.PartitionData>> entry: leaders.entrySet()) {
+
+            final int brokerId = entry.getKey().id();
+
+            calls.add(new Call("listOffsets on broker " + brokerId, context.getDeadline(), new ConstantNodeIdProvider(brokerId)) {
+
+                @Override
+                ListOffsetRequest.Builder createRequest(int timeoutMs) {
+                    return ListOffsetRequest.Builder
+                            .forConsumer(true, context.getOptions().isolationLevel())
+                            .setTargetTimes(entry.getValue());
+                }
+
+                @Override
+                void handleResponse(AbstractResponse abstractResponse) {
+                    ListOffsetResponse response = (ListOffsetResponse) abstractResponse;
+                    for (Entry<TopicPartition, PartitionData> result : response.responseData().entrySet()) {
+
+                        KafkaFutureImpl<ListOffsetsResultInfo> future = futures.get(result.getKey());
+                        PartitionData partitionData = result.getValue();
+                        if (partitionData.error == Errors.NONE) {
+                            future.complete(new ListOffsetsResultInfo(partitionData.offset, partitionData.timestamp, partitionData.leaderEpoch));
+                        } else {
+                            future.completeExceptionally(result.getValue().error.exception());
+                        }
+                    }
+                }
+
+                @Override
+                void handleFailure(Throwable throwable) {
+                    completeAllExceptionally(futures.values(), throwable);
```

Comment:
I'm not sure I follow this logic. If we get a failure on one call, why would we fail _all_ futures? Seems like we'd just want to fail the partitions associated with this call. This seems to also be a bug in `deleteRecords`.

By the way, I was looking for unit test coverage for these new APIs and didn't find any. I think we will need some test cases before we can merge.

Reply:
Good catch! Yes we only want to fail futures for `TopicPartition`s on that node. Fixing `deleteRecords` too.

Reply:
@mimaison I opened a separate PR for `deleteRecords` with some additional test cases: https://github.com/apache/kafka/pull/7449/files.